File: FoodRegistry/src/Receipt.cpp
#include "Receipt.h"
#include <iostream>
#include <iomanip>
using namespace std;

void Receipt::printSeparator(int length) {
    cout << setw(length) << setfill('=') << "" << endl;
}

void Receipt::printSeparatorToFile(int length) {
    outFile << setw(length) << setfill('=') << "" << endl;
}

void Receipt::printHeader() {
    string header = "= * === * === * === * === * === * ==";
    string title = "H & S Restaurant Meal Billing System";
    cout << header << endl << title << endl << header << endl << endl;
}

void Receipt::printHeaderToFile() {
    string header = "= * === * === * === * === * ==";
    string title = "H & S Restaurant Meal Billing System";
    outFile << header << endl << title << endl << header << endl << endl;
}

void Receipt::printCheckout(const Order& order, float amountPaid, int orderNumber) {
    float change = amountPaid - order.getTotal();

    printSeparator(36);
    cout << " Subtotal: RM " << fixed << setprecision(2) << order.getSubtotal() << endl;
    cout << " SST charge 10%: RM " << order.getSST() << endl;
    cout << " Grand total: RM " << order.getTotal() << endl;
    printSeparator(36);
    cout << "Amount received: RM " << amountPaid << endl;
    if (change > 0) {
        cout << " Change given: RM " << change << endl;
    }
    printSeparator(36);
    cout << "Thank you and have a nice day ahead!" << endl;
}

void Receipt::saveToFile(const Order& order, const vector<MenuItem>& menu, 
                        float amountPaid, int orderNumber) {
    outFile.open("ReceiptNo" + to_string(orderNumber) + ".txt");
    if (!outFile.is_open()) return;

    printHeaderToFile();
    outFile << "Order: #" << orderNumber << endl << endl;
    outFile << "Items in bill:" << endl;

    for (size_t i = 0; i < menu.size(); i++) {
        if (order.getQuantity(i) > 0) {
            outFile << "(" << menu[i].getCode() << ") [RM " 
                   << fixed << setprecision(2) << menu[i].getPrice() << "] "
                   << menu[i].getName() << " × " << order.getQuantity(i)
                   << " = RM " << menu[i].getPrice() * order.getQuantity(i) << endl;
        }
    }

    float change = amountPaid - order.getTotal();
    printSeparatorToFile(36);
    outFile << " Subtotal: RM " << fixed << setprecision(2) << order.getSubtotal() << endl;
    outFile << " SST charge 10%: RM " << order.getSST() << endl;
    outFile << " Grand total: RM " << order.getTotal() << endl;
    printSeparatorToFile(36);
    outFile << "Amount received: RM " << amountPaid << endl;
    if (change > 0) {
        outFile << " Change given: RM " << change << endl;
    }
    printSeparatorToFile(36);
    outFile << "Thank you and have a nice day ahead!" << endl;

    outFile.close();
}

File: java_src/com/foodregistry/Receipt.java
package com.foodregistry;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.List;

public class Receipt {
    
    public void saveToFile(Order order, List<MenuItem> menu, float amountPaid, int orderNumber) {
        try (PrintWriter outFile = new PrintWriter(new FileWriter("ReceiptNo" + orderNumber + ".txt"))) {
            printHeaderToFile(outFile);
            outFile.println("Order: #" + orderNumber);
            outFile.println();
            outFile.println("Items in bill:");

            for (int i = 0; i < menu.size(); i++) {
                if (order.getQuantity(i) > 0) {
                    outFile.printf("(%c) [RM %.2f] %s × %d = RM %.2f%n",
                            menu.get(i).getCode(),
                            menu.get(i).getPrice(),
                            menu.get(i).getName(),
                            order.getQuantity(i),
                            menu.get(i).getPrice() * order.getQuantity(i));
                }
            }

            float change = amountPaid - order.getTotal();
            printSeparatorToFile(outFile, 36);
            outFile.printf(" Subtotal: RM %.2f%n", order.getSubtotal());
            outFile.printf(" SST charge 10%%: RM %.2f%n", order.getSST());
            outFile.printf(" Grand total: RM %.2f%n", order.getTotal());
            printSeparatorToFile(outFile, 36);
            outFile.printf("Amount received: RM %.2f%n", amountPaid);
            if (change > 0) {
                outFile.printf(" Change given: RM %.2f%n", change);
            }
            printSeparatorToFile(outFile, 36);
            outFile.println("Thank you and have a nice day ahead!");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void printSeparatorToFile(PrintWriter outFile, int length) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < length; i++) sb.append("=");
        outFile.println(sb.toString());
    }

    private void printHeaderToFile(PrintWriter outFile) {
        String header = "= * === * === * === * === * ==";
        String title = "H & S Restaurant Meal Billing System";
        outFile.println(header);
        outFile.println(title);
        outFile.println(header);
        outFile.println();
    }
}
